batches.go

```diff
@@ -16,66 +16,57 @@ package pinecone
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"strings"
+	"text/template"
 
 	sdk "github.com/conduitio/conduit-connector-sdk"
+	cmap "github.com/orcaman/concurrent-map/v2"
 	"github.com/pinecone-io/go-pinecone/pinecone"
 )
 
-// buildBatches processes a slice of records and groups them into batches based
-// on their operation type. New batches are started whenever the operation type
-// switches from upsert to delete or vice versa.
-// Records are batched this way so that we preserve conduit's requirement of writing
-// records sequentially.
-func buildBatches(records []sdk.Record) ([]recordBatch, error) {
-	var batches []recordBatch
-	var currUpsertBatch upsertBatch
-	var currDeleteBatch deleteBatch
-
-	for i, rec := range records {
-		isLast := i == len(records)-1
-		switch rec.Operation {
-		case sdk.OperationCreate, sdk.OperationUpdate, sdk.OperationSnapshot:
-			vec, err := parsePineconeVector(rec)
-			if err != nil {
-				return nil, err
-			}
-
-			currUpsertBatch.vectors = append(currUpsertBatch.vectors, vec)
-
-			if isLast {
-				batches = append(batches, currUpsertBatch)
-			}
-			if len(currDeleteBatch.ids) != 0 {
-				batches = append(batches, currDeleteBatch)
-				currDeleteBatch = deleteBatch{}
-			}
-		case sdk.OperationDelete:
-			id := vectorID(rec.Key)
-			currDeleteBatch.ids = append(currDeleteBatch.ids, id)
-
-			if isLast {
-				batches = append(batches, currDeleteBatch)
-			}
-			if len(currUpsertBatch.vectors) != 0 {
-				batches = append(batches, currUpsertBatch)
-				currUpsertBatch = upsertBatch{}
-			}
-		}
-	}
+type recordBatch interface {
+	getNamespace() string
 
-	return batches, nil
-}
+	// isCompatible examines the given record and returns whether the
+	// record can be added to the batch or not.
+	isCompatible(sdk.Record) bool
```

Comment:
Looks like this method actually checks only for compatibility on operation. I'd suggest renaming this method to something like `isOperationCompatible`.